### DISCUSSION TITLE
Feature request/suggestion ideas

### DISCUSSION THREAD

**Discussion by greghope667** [greghope667/square-off-remote#3] [Ideas]
What features would people actually want to use? What software would people want to integrate with?

**Comment by greghope667** [comment]
Current TODOs:
- Human vs AI games
- Start position from FEN
- Reset board to initial position

Potentials:
- Integrate with some other chess software (what chess programs have appropriate APIs?)
- Lichess board API connection?

**Comment by eqikkwkp25-cyber** [comment]
## Idea: Replace PC setup for square-off-remote execution with small factor SBC setup

### My Raspberry Pi setup 

- Pisugar battery including realtime clock
- Pi Zero 2W 
- Microphone/loudspeaker phat  on top of the latter
- Epaper touch screen phat on top of the latter
- Everything within a 3D printed shell 
- Or alternatively everything injected into the GKS with the use of a Dremel and powered by the GKS battery itself?!

### Links

Kudos to [PiSugar](https://github.com/PiSugar/PiSugar)

### Photos

The device has the usual dimensions of a Pi Zero, the height matches more or less exactly the height of the GKS :-) In the second photo i put the Epaper touch screen on top of the GKS just for fun. 


![1](https://github.com/greghope667/square-off-remote/assets/73107516/b6d89241-e854-4874-97d5-8b41fa880beb)

![2](https://github.com/greghope667/square-off-remote/assets/73107516/1ab99c9f-afa9-4182-b774-a5f0cd6274e5)

**Reply by greghope667** [reply]
That setup looks awesome! I think I'm going to have to buy myself a Pi Zero + case as well once I get this all working. Personally I'm a bit reluctant to open my board up/dremel it out (I'm much more a software person really) - but it would look pretty good to have the Pi+screen in there.

**Reply by eqikkwkp25-cyber** [reply]
Thanks. You are right in the sense of focusing on the open software points, same holds for me. Hardware stuff like this is secondary and i will not use a dremel right away. I might open the GKS and apply some lubricating grease. Let me take some notes for that here.

**Reply by eqikkwkp25-cyber** [reply]
Misfortune rarely comes alone. The Epaper Touch device also broke, probably much earlier as it was not in use. Black has faded massively. A replacement device is on its way. 

Menus are certainly limited on the 250x122 screen. I may switch to the following display:
![grafik](https://github.com/greghope667/square-off-remote/assets/73107516/08d21453-0bc0-41e4-8d1a-d2d308c8fc90)

**Comment by eqikkwkp25-cyber** [comment]
## Gimmick or not: Voice Control

I would like to discuss this here: Would we like to control every move via voice or just starting games, choosing game options, stopping games ...? Are those features just proof of concept or could they be really useful?

Compare Coqui STT project [3d-voice-chess](https://github.com/HarikalarKutusu/3d-voice-chess)

My guess is that voice.py does not work properly, i.e. the WER is too high (bad accuracy). This can be resolved  by manual modification of the PS model (restricting dictionary and language model to a minimum of words used in voice.py and).

**Reply by greghope667** [reply]
The current voice control does work for me at least (but requires a very quiet room and decent microphone). How well it works is going to depend a lot on the user I guess. It's a pretty simple engine though that can be run locally which I think is worth keeping in for now. There are a bunch of parameters that can be tweaked, it could be worth looking into changing those.

If we wanted to get more complex I'd like to see if there are any speech-to-chess projects I can pull in as a library. There seem to be a few similar projects about and it would be nice to cooperate with those.

**Reply by eqikkwkp25-cyber** [reply]
Yes, Pocketsphinx will be sufficient for this job. One needs to restrict the potential scope of words, i.e. the delivered dictionary and language model. As a result the accuracy will be much better. I will have a look into this later. 

I own different kind of micarrays for the Raspberry Pi: 8mic, 4mic, 2mic and a simple USB 1mic. For sure more is better in some sense, i.e. for far distance recognition. You can whisper from 5 meters away towards the 8mic one and it will recognize the spoken words. For our purpose a simple 1mic will do, already tested it in the past for other projects.

**Comment by eqikkwkp25-cyber** [comment]
## Minimizing noise that comes from the motor

As shown in this video [What's Inside a Square Off Chess Board and where is Intelligence Artificial module ...](https://yewtu.be/watch?v=g9kKoZJoH7o&t=261) one could use **lubricating grease** to fix some of the noise. 

The video shows how to open the device (Kingdom Set) and some details are visible (SBC, stepping motor NEMA17HS0410 etc). Its clear how the Kingdom Sets work, similar to "**x-y-writer**s" which are used for 3D printers. Neo and Swap seem to have a robot arm. 

Some 3D printers are modified using step motor dampers for noise reduction, i think it does not makes sense for the (G)KS, compare the 3D printer modification [Voltlog #181 - Installing Stepper Motor Vibration Dampers On The CR10](https://yewtu.be/watch?v=GOknoEx1fWg)


### Photos

![overview](https://github.com/greghope667/square-off-remote/assets/73107516/29a78a19-4cbd-4fe9-a08e-45f8281aec85)

![grease](https://github.com/greghope667/square-off-remote/assets/73107516/5e3e1323-a13c-4fc4-8991-d5c722833957)


![board](https://github.com/greghope667/square-off-remote/assets/73107516/e74b8597-f461-411b-a8d1-76ca27580cea)

![motor](https://github.com/greghope667/square-off-remote/assets/73107516/5dc9ad01-aa8e-43b6-baa5-10c9943f1de9)

**Reply by eqikkwkp25-cyber** [reply]
To make a long story short: I did it, i.e. opened the device. Would I recommend it to others? Certainly not because of the noise reduction. Grease was already present and renewed lubrication brought about -4 decibels. Tested with an app and the following game start

`[srcYtaIY]  -- Stockholm Team Championship 2023/24 -- September 18th - January 29th 
[jYV3LuDH] -- Round 3 -- September 18th - January 29th 
[Round 3: Tomas Karlsson - Jung Min Seo Tomas Karlsson vs Jung Min Seo] -- *      
e2e4 c7c5
g1f3 d7d6
d2d4 c5d4

before (outside measurement): max 33 eq 22 min -5
after (outside measurement): max 29 eq 18 min -2`

If you want to open your device for any reason (at your own risk), here are a few tips:
1) There are three symmetrical screws on each side and four asymmetrical screws on the top and bottom. Unless Squareoff has left one or the other out :-)
2) You can feel these and unscrew them directly through the velvet. 
3) Avoid the screws of the two motors.
4) When opening, pay particular attention to the flex cable that connects the chessboard and the Bluetooth board. In my case, this was glued with hot glue :-) The on/off switch and charging socket are easier to handle. 

It may well happen that your device calibrates suboptimally after the procedure. Seems to work properly after some iterations.

**Reply by eqikkwkp25-cyber** [reply]
I am not sure with regards to the electromagnet, that looks like wear and tear to me, compare wit the above one, red material is velvet

![7](https://github.com/greghope667/square-off-remote/assets/73107516/4faeaf90-0843-4fcb-96c1-3b3ec23ae0ec)

**Reply by eqikkwkp25-cyber** [reply]
### The ugly, the bad, the good

## The ugly
I'm back in the game after about a week :-) 

I must have opened the device a dozen times to make initial optimizations regarding noise development and later regarding suboptimal centering of the pieces (not centered) after about 50-60 moves...

... until the board, more precisely the touch function during moves and the mechanical dragging, stopped working. Hardware problem with flex cable was my assumption until I used the Python version instead of the Android SquareOff app and voila ... In any case, I haven't understood it (yet). 

## The bad
There is hardly any suitable space to install a Raspberry Pi in the GKS, for example. The issue is off the table.

## The good
Grease applied, electromagnet velvet replaced, silicone dampers placed on electric motors and such to absorb the shocks at the ends of the board.

**Comment by eqikkwkp25-cyber** [comment]
## Considering alternative  chess engines

Stockfish engine with lowest skill level could be hard to beat for beginners, i guess i read its about 1300 ELO but it seems to play better. As a proof of concept i added maia1 (reference [maia-chess](https://github.com/CSSLab/maia-chess)) to player.py using a wrapper shell script which starts lc0 (reference [lczero](https://lczero.org/)) with maia1 model as weights. Maia1 is about 1100 ELO and should play more human like.  Moves are deterministic but you can modify UCI parameters so that the engine plays random moves. 

Could be difficult to deliver this to the repository as commands and locations will differ between operating systems and might be harder to install for some users. Compilation of lc0 on Linux is straightforward.

**Reply by greghope667** [reply]
Stockfish, even on low difficulties, is surprisingly hard to beat - I think due to it playing 'computer-like' moves. Lowering the search depth/node count helps a bit here but is tricky to get right. Maia might be much more fun to play against.

The best way to talk to engines is as a separate process through the UCI or XBoard protocols (like how you added Maia). I think some sort of configuration UI for engines would be helpful - similar to how chess GUIs work. Then it would be fairly easy to add/configure engines for users.

**Reply by eqikkwkp25-cyber** [reply]
That is correct. For your information: square-off-remote + Stockfish is running smooth with on a Raspberry Pi Zero 2.  Playing against other engines via Lichess API could be an alternative (all Maia models are up an running there) to avoid this kind of underlying architecture issue.

**Reply by greghope667** [reply]
I'm going to look at connecting to the Lichess [board api](https://lichess.org/api#tag/Board), it's designed for exactly this use case of connecting an external board. I'd still like to work on running engines locally at some point, I'll need to think about how best to do that.

**Reply by greghope667** [reply]
I have initial support for the board api - enough to be able to play online at least - so you should be able to play against Maia (or other bots) online.

**Reply by eqikkwkp25-cyber** [reply]
I'm back, details are above. Great that you implemented this so quickly. Just tried it briefly and the code doesn't work for me. Maybe I'm doing something wrong? I will open a ticket.

p.s. Works now. I guess i expected that it starts automatically some game. After challenging a bot it worked out and the game was synced with the GKS board. Great.

**Comment by greghope667** [comment]
# User interface

Now that basic functionality is implemented, I want to add a better interface so we don't have to manually run shell commands. I'm considering adding a simple GUI, with commands (to start with):
- Start local game
- Select opponent/configure engines
- Start online game
- Watch online games

**Reply by eqikkwkp25-cyber** [reply]
Both todos 

- >     Start position from FEN
- >     Reset board to initial position

have lower prio for me now. 

Puzzles are not in my focus anymore, would take to much time to setup a FEN. And resetting the pieces manually is a kind of meditation now. 

I guess a GUI would be the most wanted feature for all non-programmers / end-users.  I guess we still can use features from the command line too, its just about putting things together...

**Comment by eqikkwkp25-cyber** [comment]
### Chess coaching, evaluations etc. 

It will be harder to realize but i saw [How to play ChessUp](https://yewtu.be/watch?v=onkmw11Xngg&t=106) and thought "Damn, no LEDs ...". After a while a realized that we could return illegal moves from position A to position A and computer evaluation of those as "blunder", "mistake", "inacurracy". Output could be on screen, via voice, blinking LEDs etc. 

Same holds for moves which have been already done.
